### PR TITLE
Addition of BigtableDataGCJClient with GCJ adapters

### PR DESCRIPTION
**Pull request by rahulKQL**
- Apart from [AbstractTestFilter#testInterleaveNoDuplicateCells](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java#L2011-L2033) all integration test case are passing for both H1 & H2. Currently marked that test case as ignored will update once resolved.
- Now returning QualifierFilter instance instead of ChainFilter in ColumnPrefixFilterAdapter.java. (Might have mistakenly returned ChainFilter).
- Added RowResultScanner to translate ServerStream into ResultScanner(will add test case).
- Have created StreamObserverAdapter similar to ApiStreamObserver from GCJ.

Test Cases can be invoked with:
```
mvn clean verify \
       -PbigtableIntegrationTest \
       -Dgoogle.bigtable.project.id=grass-clump-479 \
       -Dgoogle.bigtable.instance.id=node-client-performance \
       -Dgoogle.bigtable.use.gcj.client=true
```

Note: The GCJ returns duplicate values when using multiple filters with `interleave`, If the same set of filters used with Filters#chain none it works fine.

To demonstrate duplication from `Filter#interleave`, I have added one example in [`TestBigtableVeneerClientFactory`](https://github.com/googleapis/cloud-bigtable-client/compare/master...rahulKQL:gcj_data_client_impl#diff-a9e488e7dc2d8858b2c25e7a08c272a0R194).

### CODE REVIEW

**Comment by rahulKQL** [comment]
@sduskis,

For the [testInterleaveNoDuplicateCells](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java#L2011-L2033) issue I tried discarding Cells with labels from FlatRowAdapter but doing that makes [WhileMatchFilter allow every cell](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/BigtableWhileMatchResultScannerAdapter.java#L116-L137). 
Then I tried [adding labels into cell comparison](https://github.com/googleapis/cloud-bigtable-client/pull/2128/files#diff-e2e90efec4bfb8d9914f43602303a3dfR171), that works as expected. 

As of now, I am not sure about performance overhead with comparing labels of each cell (_To keep performance in check I only checked for the difference in size of the labels as order does not matter to us_).

There is one more approach( to check each cell with labels with "-in" & "-out" and discard others. But that sounds more verbose.

Please have a review and let me know your thought on this.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGCJClient.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  @Override
+  public ResultScanner<FlatRow> readFlatRows(Query request) {
+    final ServerStream<FlatRow> stream =
+        delegate.readRowsCallable(new FlatRowAdapter()).call(request);
+    return new RowResultScanner<>(stream, new FlatRow[0]);
+  }
+
+  private ResultScanner<FlatRow> readFlatRowsV2(Query request){
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGCJClient.java]: Is `readFlatRowsV2` required?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGCJClient.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  @Override
+  public ApiFuture<List<FlatRow>> readFlatRowsAsync(Query request) {
+    return delegate.readRowsCallable(new FlatRowAdapter()).all().futureCall(request);
+  }
+
+  @Override
+  public void readFlatRowsAsync(Query request, StreamObserver<FlatRow> observer) {
+    //TODO: figure out to convert StreamObserver to ResponseObserver.
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGCJClient.java]: Is this `//TODO` still valid?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 3]
```diff
@@ -250,6 +253,10 @@ public BigtableSession(BigtableOptions opts) throws IOException {
     dataClient =
         new BigtableDataGrpcClient(dataChannel, sharedPools.getRetryExecutor(), options);
     dataClient.setCallOptionsFactory(callOptionsFactory);
+    BigtableDataSettings dataSettings =
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: Please delay creating the `dataSettings` and `dataGCJClient` until the client is actually needed.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 4]
```diff
@@ -351,6 +358,9 @@ public RequestContext getDataRequestContext() {
    * @return a {@link IBigtableDataClient} object.
    */
   public IBigtableDataClient getClientWrapper() {
+    if(options.useGCJClient()){
+      return dataGCJClient;
+    }
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: Please use `else` here.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java [hunk 5]
```diff
@@ -102,10 +106,9 @@ public void cellValue(ByteString value) {
     public void finishCell() {
       if (!Objects.equals(this.family, this.previousFamily)) {
         previousFamily = this.family;
-        currentFamilyCells = new ArrayList<>();
+        currentFamilyCells = new TreeSet<>(compareCells());
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: This should really be a List, and not a `TreeSet`

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java [hunk 6]
```diff
@@ -121,8 +124,9 @@ public void finishCell() {
      */
     @Override
     public FlatRow finishRow() {
-      ImmutableList.Builder<FlatRow.Cell> combined = ImmutableList.builderWithExpectedSize(totalCellCount);
-      for (List<FlatRow.Cell> familyCellList : cells.values()) {
+      ImmutableList.Builder<FlatRow.Cell> combined =
+          ImmutableList.builderWithExpectedSize(totalCellCount);
+      for (Set<FlatRow.Cell> familyCellList : cells.values()) {
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: Please compare the current and previous for the following:

```
FlatRow.Cell previous;
for(FlatRow.Cell c : cells.values) {
  if(previous == null || !c.getLables().isEmpty() || !keysMatch(c, previous)){
    combined.add(c);
  }
}
```

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java [hunk 7]
```diff
@@ -133,8 +134,15 @@ private void initializeClients() throws IOException{
   }
 
   /**
-   * This test runs only if it finds "test.client.project.id" & "test.client.project.id"
-   * VM arguments. Then it calls to an actual Bigtable Table & performs the checks below:
+   * This test runs only if it founds below env variables,
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java]: s/founds below env variables/finds the environment variables listed below`

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java [hunk 8]
```diff
@@ -133,8 +134,15 @@ private void initializeClients() throws IOException{
...
+   * Then it calls to an actual Bigtable Table & performs the checks below:
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java]: Please don't use `&`, since the comments are valid HTML.  Please use the word `and`

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java [hunk 9]
```diff
@@ -183,20 +191,35 @@ public void testWithActualTables() throws Exception{
       dataClient.mutateRow(rowMutation);
       LOG.info("Successfully Mutated");
 
-      Query query = Query.create(TABLE_ID);
+      Query query = Query.create(TABLE_ID)
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/config/TestBigtableVeneerSettingsFactory.java]: This seems to be part of a separate PR.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java [hunk 10]
```diff
@@ -2009,6 +2009,7 @@ private static void assertNextEquals(ResultScanner scanner, byte expectedKey[])
   }
 
   @Test
+  //TODO(rahulkql): failing due to duplicate Cells when used with interleave
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Please remove this comment.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/ColumnPrefixFilterAdapter.java [hunk 11]
```diff
@@ -45,9 +45,8 @@ public Filter adapt(FilterAdapterContext context, ColumnPrefixFilter filter)
     ByteString.Output outputStream = ByteString.newOutput(prefix.length * 2);
     ReaderExpressionHelper.writeQuotedExpression(outputStream, prefix);
     outputStream.write(ReaderExpressionHelper.ALL_QUALIFIERS_BYTES);
-    Filter qualifierFilter = FILTERS.qualifier().regex(outputStream.toByteString());
 
-    return FILTERS.chain().filter(qualifierFilter);
+    return FILTERS.qualifier().regex(outputStream.toByteString());
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/ColumnPrefixFilterAdapter.java]: Please make this change in a different PR.

File: bigtable-hbase-1.x-parent/bigtable-hbase-1.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestFilters.java [hunk 12]
```diff
@@ -37,7 +37,7 @@
 import static com.google.cloud.bigtable.hbase.test_env.SharedTestEnvRule.COLUMN_FAMILY;
 
 public class TestFilters extends AbstractTestFilters {
-
```
Comment (sduskis) [inline bigtable-hbase-1.x-parent/bigtable-hbase-1.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestFilters.java]: Please make this change in a different PR.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 13]
```diff
@@ -265,10 +268,18 @@ public BigtableSession(BigtableOptions opts) throws IOException {
 
     BigtableClientMetrics.counter(MetricLevel.Info, "sessions.active").inc();
 
-    // Defer the creation of both the tableAdminClient until we need them.
-    //TODO(rahulkql): Need to identify which resources to initialized for GCJ's adapter case.
-    this.adminSettings = BigtableVeneerSettingsFactory.createTableAdminSettings(options);
-    this.baseAdminSettings = BaseBigtableTableAdminSettings.create(adminSettings.getStubSettings());
+
+    //TODO(rahulkql): Need to identify which resources to initialized for GCJ's adapter case
+    if(options.useGCJClient()){
```
Comment (rahulKQL) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: @sduskis 
Have included `dataSettings` & `dataClient` creation in xor of `BigtableSession` only in case of `options.useGCJClient.` 
As moving it to getClientWrapper would introduce a chain of updates to handle `IOException`.

If needed, I could create a separate PR on which we can move these into `BigtableSession#getClientWrapper` & `BigtableSession#getAdminClientWrapper`.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java [hunk 14]
```diff
@@ -122,13 +122,23 @@ public void finishCell() {
...
+          }
+          previous = c;
+        }
       }
 
       return new FlatRow(this.currentKey, combined.build());
     }
 
+    private boolean keysMatch(Cell c, Cell previous) {
+      return c.equals(previous);
```
Comment (rahulKQL) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: @sduskis 
I had a doubt here, Using Cell#equals would be ok? Or shall I compare specific field i.e. labels only?

Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: Please only use the qualifier and timestamp, since the row key and family are definitely going to be the same.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java (hunk 13 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: Please wrap the `IOException` in a `RuntimeException` for now.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java (hunk 14 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: I have updated to compare with qualifier & timestamp... but it seems labels comparison are needed for the case of `WhileMatchFilter`.

**Comment by rahulKQL** [comment]
@sduskis 

Thanks for the suggestions, I have made update Please have a look

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java [hunk 15]
```diff
@@ -121,14 +121,27 @@ public void finishCell() {
...
+        }
       }
 
       return new FlatRow(this.currentKey, combined.build());
     }
 
+    private boolean keysMatch(Cell current, Cell previous) {
+      return current.getTimestamp() == previous.getTimestamp()
+          && Objects.equals(current.getQualifier(), previous.getQualifier())
+          && Objects.equals(current.getLabels(), previous.getLabels());
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java]: Please remove the `getLabels()` comparison.

**Comment by sduskis** [comment]
@rahulKQL, if you think that this PR is ready, please remove "WIP: " from the title.

**Comment by rahulKQL** [comment]
@sduskis 

Thanks for pointing it out, This PR is ready for another round of review.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        break;
+      }
+      resultList.add(row);
+    }
+    return resultList.toArray(arr);
+  }
+
+  @Override
+  public int available() {
+    throw new UnsupportedOperationException("this operation is not supported");
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java]: @igorbernstein2, is there anyway to get the in-memory count of available rows?

Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java]: No, the  gcj client only parses rows on demand. Closest you can do is `iterator.hasNext() ? 1 : 0;`
Remind me how this is used?

Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java]: `ResultScanner#availble` returns the number of rows available without blocking the network. 
Should we use `stream.isReceiveReady() ? 1 : 0` ? as this will be a non-blocking call.
@sduskis 
Please let me know if we should return a value or continue to throw an exception.

Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java]: Please return a value.

Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowResultScanner.java]: Have updated this method with `stream.isReceiveReady() ? 1 : 0`
